reader.go

```
@@ -0,0 +1,125 @@
+package main
+
+import (
+	"io"
+	"os"
+
+	"github.com/saintfish/chardet"
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/encoding/charmap"
+	"golang.org/x/text/encoding/japanese"
+	"golang.org/x/text/encoding/korean"
+	"golang.org/x/text/encoding/simplifiedchinese"
+	"golang.org/x/text/encoding/traditionalchinese"
+	"golang.org/x/text/encoding/unicode"
+	"golang.org/x/text/transform"
+)
+
+const (
+	sampleSize = 8192 // Increased sample size for better accuracy
+)
+
+func detectEncoding(path string) (encoding.Encoding, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	sample := make([]byte, sampleSize)
+	n, err := io.ReadFull(f, sample)
+	if err != nil && err != io.ErrUnexpectedEOF && err != io.EOF {
+		return nil, err
+	}
+	sample = sample[:n]
+
+	detector := chardet.NewTextDetector()
+	result, err := detector.DetectBest(sample)
+	if err != nil {
+		return nil, err
+	}
+
+	switch result.Charset {
+	case "UTF-8":
+		return unicode.UTF8, nil
+	case "UTF-16BE":
+		return unicode.UTF16(unicode.BigEndian, unicode.IgnoreBOM), nil
+	case "UTF-16LE":
+		return unicode.UTF16(unicode.LittleEndian, unicode.IgnoreBOM), nil
+	case "UTF-32BE":
+		// Go's x/text doesn't directly support UTF-32, fallback to UTF-8

```

Silently falling back to `UTF-8` seems wrong to me. This should at least be logged.
```go
log.Printf("unsupported encoding: %s, falling back to UTF-8", result.Charset)
```

Same goes for unknown encodings (I also don't get why you return nil here instead of the `UTF-8` fallback).
```go
log.Printf("unknown charset: %s", result.Charset)
```

I think you should log errors and encodings in general.
```go
if err != nil {
    log.Printf("error detecting encoding: %s", err)
    ...
} else {
    log.Printf("encoding detected: %s", enc)
}
```